app/uk/gov/hmrc/cgtpropertydisposals/repositories/CacheRepository.scala

```diff
@@ -71,18 +86,74 @@ trait CacheRepository[A] {
         }
     }
 
-  private def dropInvalidIndexes: Future[Unit] =
+  def findAll(ids: List[String]): Future[Either[Error, List[A]]] =
+    collection
+      .find(
+        Json.obj("_id" -> Json.obj("$in" -> ids)),
+        None
+      )
+      .cursor[JsObject]()
+      .collect[List](Int.MaxValue, Cursor.FailOnError[List[JsObject]]())
+      .map { list =>
+        val (errors, values) = EitherT(list.map(readJson))
+          .subflatMap(
+            Either.fromOption(_, "Could not find value in JSON")
+          )
+          .value
+          .partitionWith(identity)
+
+        if (errors.nonEmpty)
+          Left(Error(s"Could not get all values: ${errors.mkString(", ")}]"))
+        else
+          Right(values)
+      }
+      .recover {
+        case exception => Left(Error(exception))
+      }
+
+  def find(id: String): Future[Either[Error, Option[A]]] =
+    collection
+      .find(
+        Json.obj("_id" -> id),
+        None
+      )
+      .one[JsObject]
+      .map {
+        case None       => Left(Error(s"Could not find json for id $id"))
+        case Some(json) => readJson(json).leftMap(Error(_))
+      }
+      .recover {
+        case exception => Left(Error(exception))
+      }
+
+  private def readJson(jsObject: JsObject): Either[String, Option[A]] =
+    (jsObject \ objName)
+      .validateOpt[A]
+      .asEither
+      .leftMap(e ⇒ s"Could not parse session data from mongo: ${e.mkString("; ")}")
+
+  private def dropInvalidIndexes(): Future[Unit] =
     collection.indexesManager.list().flatMap { indexes =>
       indexes
         .find { index =>
-          index.name.contains(indexName) &&
+          index.name.contains(cacheTtlIndexName) &&
           !index.options.getAs[Long]("expireAfterSeconds").contains(cacheTtl.toSeconds)
         }
         .map { i =>
           logger.warn(s"dropping $i as ttl value is incorrect for index")
-          collection.indexesManager.drop(indexName).map(_ => ())
+          collection.indexesManager.drop(cacheTtlIndexName).map(_ => ())
         }
         .getOrElse(Future.successful(()))
     }
 
+  private def toJodaDateTime(dateTime: LocalDateTime): org.joda.time.DateTime =
```

Comment:
why do we need this?

Reply:
the TTL depends on there being a valid date type in document somewhere - in the insert method where we have:
```scala
val lastUpdated: DateTime = overrideLastUpdatedTime.map(toJodaDateTime).getOrElse(time)

val modifier = Json.obj(
        "$set" -> Json
          .obj(
            objName       -> Json.toJson(value),
            "lastUpdated" -> lastUpdated
          )
      )
```
the thing inside `Json.obj` expects a sequence of `(String,JsValueWrapper)`. There only seems to be implicit conversions from joda time classes and not java time classes that are supplied out of the box. Hence, the conversion to joda.